src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorCodeLensProvider.ts

```diff
@@ -0,0 +1,130 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import { isRazorCSharpFile, getRazorDocumentUri } from './RazorConventions';
+import { RazorLanguageFeatureBase } from './RazorLanguageFeatureBase';
+import { LanguageKind } from './RPC/LanguageKind';
+import { CancellationToken } from 'vscode-jsonrpc';
+
+class RazorCodeLens extends vscode.CodeLens {
+    constructor(
+        range: vscode.Range,
+        public uri: vscode.Uri,
+        public document: vscode.TextDocument,
+        command?: vscode.Command) {
+
+        super(range, command);
+    }
+}
+
+export class RazorCodeLensProvider
+    extends RazorLanguageFeatureBase
+    implements vscode.CodeLensProvider {
+
+    public async provideCodeLenses(document: vscode.TextDocument, token: vscode.CancellationToken) {
+        try {
+            const razorDocument = await this.documentManager.getDocument(document.uri);
+            if (!razorDocument) {
+                return;
+            }
+
+            const csharpDocument = razorDocument.csharpDocument;
+
+            // Get all the code lenses that applies to our projected C# document.
+            const codeLenses = await vscode.commands.executeCommand<vscode.CodeLens[]>(
+                'vscode.executeCodeLensProvider',
+                csharpDocument.uri) as vscode.CodeLens[];
+
+            // Re-map the CodeLens locations to the original Razor document.
+            const remappedCodeLenses = new Array<vscode.CodeLens>();
+            for (const codeLens of codeLenses) {
+                const result = await this.serviceClient.mapToDocumentRange(
+                    LanguageKind.CSharp,
+                    codeLens.range,
+                    razorDocument.uri);
+                if (result) {
+                    const newCodeLens = new RazorCodeLens(result.range, razorDocument.uri, document, codeLens.command);
+                    remappedCodeLenses.push(newCodeLens);
+                } else {
+                    // This means this CodeLens was for non-user code. We can safely ignore those.
+                }
+            }
+
+            return remappedCodeLenses;
+
+        } catch (error) {
+            return [];
```

Comment:
Hmm are any errors here expected? Wondering if we should be logging the error via `RazorLogger`

Reply:
Nothing specific to expect. I was just being defensive because this method is called every time a Razor file loads/changes.

Reply:
I second a call to `RazorLogger` here. In general I'd rather be too loud about errors than too quiet.